### ISSUE TITLE
Feature to track multicore compatibility of libraries

### ISSUE THREAD

**Issue by gadmm** [ocaml/opam#6194]
In order to be used in a multicore setting, libraries typically need to be written with multicore in mind. OCaml 4 libraries must be adapted to multicore in various ways (e.g. eliminating reliance on a global lock, adjusting to possible new FFI guidelines, etc.). OCaml 4 libraries should be assumed non-muticore-compatible, unless the authors indicate explicitly that thought and work has been put into ensuring it.

The tooling is currently missing in the ecosystem for keeping track of multicore-readiness of libraries. I propose to discuss this problem here. For instance, there could be some kind of tagging in opam to track library capabilities, with a "multicore" tag that is absent by default.

cc @kayceesrk

**Comment by hannesm** [comment]
Thanks for your constructive proposal, I had a similar question at the opam-repository issue tracker some time ago (and was surprised by the answer) https://github.com/ocaml/opam-repository/issues/25524

**Comment by avsm** [comment]
Why was my answer surprising, @hannesm? What I described there still seems broadly correct to me. I think the way to realistically document parallellism-safety isn't at the opam level, but at the odoc level. Given that we now have documentation generation centrally, it's up to library authors to document what locks their OCaml interfaces require -- and nothing is stopping that as far as I can see from a technical perspective.

**Comment by dbuenzli** [comment]
> OCaml 4 libraries should be assumed non-muticore-compatible, unless the authors indicate explicitly that thought and work has been put into ensuring it.

That position sounds extreme to me. In OCaml 4 you already had to mention if your libraries were thread-safe or not (e.g. grepping through my code bases [here](https://github.com/dbuenzli/bos/blob/cead0b88fb846463574201e343c3fab9df7c5d5a/src/bos.mli#L396), [here](https://erratique.ch/software/react/doc/React/index.html#update) or [there](https://github.com/dbuenzli/rel/blob/3deee04b1f70e0a341a2baed6b0dee5f637d894b/src/sqlite3/rel_sqlite3.mli#L8-L12)). 

I would take the opposite stance, which is the one that was adopted upstream assume. Assume thread-safety and document unsafeties. You can give as much "thought and work", stuff is going to slip through anyways.

**Comment by kayceesrk** [comment]
After discussing with @avsm elsewhere, I've come to a similar conclusion @dbuenzli. We haven't done anything special for thread safety in OCaml 4. I do not see a strong reason to things differently with parallelism.

**Comment by dbuenzli** [comment]
But that's not to say I would not be interested by: 

> The tooling is currently missing in the ecosystem for keeping track of multicore-readiness of libraries

:–)

**Comment by avsm** [comment]
Yeah, it would be _really_ useful to have some Merlin or LSP support for querying parallel-safety docstrings/alerts from compiled dependencies in the environment in a more structured way. I haven't had a chance to think through what that might look like concretely, though.

**Comment by dbuenzli** [comment]
Why not. 

From a library sanitization perspective I think I would rather like a simple lint tool that looks up for [non-atomic top-level refs](https://github.com/dbuenzli/cmdliner/blob/cb235b559e29a063fe3c4fb1a60b444cf70dd402/src/cmdliner_base.ml#L10-L17), parallel unsafe use of `Format` (likely to be my largest issue), etc.

Speaking of parallel unsafe use of `Format` I noticed yesterday that 10 years ago I [removed `to_string` functions from `Gg`](https://github.com/dbuenzli/gg/blob/master/CHANGES.md#v090-2014-08-23-cambridge-uk) which were using the `pp` functions with `Format.str_formatter` on the ground that they were not thread-safe…

**Comment by dra27** [comment]
opam's `tags` field could be using for this, but I'm not sure that there's anything which needs changing in opam itself to be able to do that - is there anything you have in mind that would be specifically wanted in the opam client itself?

**Comment by dra27** [comment]
Unsurprisingly, I come quite strongly down on what we did with 5.0 - that is, not blocking single-domain-only-but-not-necessarily-checked-for-multicore packages from working on 5.0 by default. Fracturing the existing single-core ecosystem as part of 5.0 would have been a big mistake, and a vast amount of effort was poured into trying not to do that.

It's easy in opam-repository to mark packages which have _not_ been tested for parallelism and then set-up an opam switch which forbids these packages. However, that brings a lot of bureaucracy for new releases of still-untested packages, so I'm, not sure it would be that useful. Sadly, there isn't as easy a way to do the converse - have a switch which can be set to _only_ install packages which have been tested.
